### ISSUE TITLE
[IMPORTANT] Upcoming Frisby v2.0

### ISSUE THREAD

**Issue by vlucas** [vlucas/frisby#316]
I have started on a new version of Frisby.js that will be released as v2.0 [in the "v2" branch of the Frisby.js project on GitHub](https://github.com/vlucas/frisby/tree/v2). Final timeline is unknown at this point, but it has very good progress, and I am already using it in some of my own personal projects with great success.

The main goal is to re-focus on what Frisby does best, give you more control over the Jasmine test structure, and to never nest Jasmine test contexts so issues like the custom matcher functions not working never happen - i.e. issues #93 #94 #307, etc.
### Changes in v2.0

With v2.0, Frisby tests will now sit **inside your own Jasmine test, and Frisby will only make HTTP requests and run assertions on the response**. This gives you a lot more control over the test itself, and gives you the ability to use Jasmine's other methods for flow control, like `beforeAll`, `beforeEach`, etc.

Example:

``` javascript
var frisby = require('frisby');

it('should get user Joe Schmoe', function(doneFn) {
  frisby.get(testHost + '/users/1')
    .expect('status', 200)
    .expect('jsonContains', {
      id: 1,
      email: 'joe.schmoe@example.com'
    })
    .done(doneFn);
});
```

I will have a lot more details later with a full rundown of all the features, but with all the stagnant issues piling up, I just wanted to give everyone a heads up on what's going on with this project.
### Using v2 Now

If you want to use Frisby.js v2 now, you can install it by targeting the GitHub v2 branch:

```
  "devDependencies": {
    "frisby": "git://github.com/vlucas/frisby.git#v2",
    "jasmine-node": "^1.14.5"
  },
```

If you want to help build the next version of Frisby, pull requests are welcome.
### What happens to v0.85?

Unfortunately, the core issues often reported here are not fixable with the current API since Frisby wraps and generates the entire Jasmine test for you. This is very limiting in a lot of ways, and nested Jasmine tests is the largest of these very limiting and unfixable issues. Additionally, the release of Jasmine 2.0 has put the release process on hold since a lot of things have changed, and the main dependency of `jasmine-node` _has still not released an official stable version for Jasmine 2.0_.

Therefore, given the poor API design, the current unfixable issues related to that, and the fact that Frisby.js is over 4 years old in its current form and stuck on Jasmine v1.0, the current release will probably be the last in the v0.x series, and the current v0.x code will be abandoned and replaced with the v2.x branch.

**Comment by dnperfors** [comment]
A quick question: basically v2 will have less connection with Jasmine, so would it be possible to use Frisby with QUnit or another testing framework as well?

**Comment by vlucas** [comment]
Yes, that is the idea. The [current provided expectations](https://github.com/vlucas/frisby/blob/v2/src/frisby/expects.js) use Node.js's built-in `assert` library, which throws exceptions on failures instead of expecting Jasmine methods like `expect` to be available, so this should indeed be possible.

**Comment by sharadJay** [comment]
@vlucas are there any changes with after and afterjson hooks ? I am not able to use them with my tests.

**Comment by vlucas** [comment]
Currently `after` and `afterJSON`  will be replaced with `then`, which will serve the same purpose, and function like promises.

**Comment by dnperfors** [comment]
@vlucas I am starting with some experiments with frisby v2 using several test engines, so I tried to write a failing test. But even using Jasmine I don't see the test failing. Am I missing something?

``` javascript
'use strict';

const frisby = require('frisby');

const testHost = 'http://httpbin.org';

describe('Frisby', function() {
    it('Test expectStatus works as... well, expected', function(doneFn) {
        frisby.fetch(testHost + '/ip')
            .expect('status', 400)
            .done(doneFn);
    });
});
````

**Comment by vlucas** [comment]
@dnperfors Yes - there are some issues with Jasmine and async tests in v2. It will fail properly if you run it with `jasmine-node`, however. I am still working on it :).

**Comment by sharadJay** [comment]
@vlucas My response returns array of offers, with earlier version I was able to use path like this for expectJSONType
```javascript
.expectJSONTypes('offers.*', {
```
but now on using with frisby 2.0 I get error
```
   Message:
     Expected 'TypeError: Expected 'offers.*' to be Array (got 'object' from JSON response)
```

Can you help with this ?

**Comment by vlucas** [comment]
Can you make a reproducible test using `frisby.fromJSON({ ... })` instead of your HTTP call so I can see what's going on?

**Comment by sharadJay** [comment]
This is test code
```javascript
var frisby = require('frisby');
var Joi = frisby.Joi;


describe("a suite", function () {
    it('should match JSON with nested structure and single path', function (doneFn) {
        frisby.fromJSON({
            "offers": [{"name": "offer1"}, {"name": "offer2"}]
        })
            .expect('jsonTypes', 'offers.*', {
                "name": Joi.number()
            })
            .done(doneFn);
    })
})
;

```

Stacktrace
```
a suite - 18 ms
    should match JSON with nested structure and single path - 17 ms

Failures:

  1) a suite should match JSON with nested structure and single path
   Message:
     Expected 'TypeError: Expected 'offers.*' to be Array (got 'object' from JSON response)
    at acceptance-tests/functional-tests/node_modules/frisby/src/frisby/utils.js:23:15

```

**Comment by vlucas** [comment]
This has been fixed - sorry for the issues here. I also added a `use` method that can be used to modify frisby settings per test in a more modular way.

**Comment by sharadJay** [comment]
@vlucas Yes thank you this works. I see that assertions in expectJson are not reported if they pass, although if checks fail this is reported in number of failed assertions.

**Comment by sharadJay** [comment]
Also I see that timeout function is missing which was there in previous version, can we add it ?

**Comment by sharadJay** [comment]
@vlucas I am not able to post data by removing json:true as mentioned in #233 It was working in previous version. Can you look at it ?

**Comment by philippesillon** [comment]
@vlucas I try to get V2 by using command 

`npm install git://github.com/vlucas/frisby.git#v2`

 I got error ENOENT. 

![image](https://cloud.githubusercontent.com/assets/5950342/17034834/5f630f36-4f52-11e6-837e-c941cd8710fc.png)

May be the syntax is not good ?

**Comment by vlucas** [comment]
Looks like a local permissions issue. The top line says "checkPermissions: Missing write access to <folder>". I was just able to install Frisby v2 locally, so it works.

**Comment by kibertoad** [comment]
Are there any ETAs for the RC version at least?

**Comment by vlucas** [comment]
There is no ETA for v2, but I hope to put an initial beta or RC release out in early August.

The current v2 code already has nearly the same amount of features as v1, but there are a few more things to do. The v2 code is also already more stable and better functioning that v1 as well, if you want to get an early start. I have multiple code bases already using Frisby v2 (so I can dogfood it), and the API won't be changing much ;)

**Comment by kibertoad** [comment]
Thank you! Will an RC be released on NPM?

**Comment by vlucas** [comment]
Yes - It will be released on NPM when the time comes.

**Comment by saurabh1gupta** [comment]
Question on v2: Will v1 still be supported? As we always have test cases implemented in v1 which we would like to run continuously without any changes (or minimal change) ?

**Comment by xxjthxx** [comment]
@vlucas Would be great if you could be more transparent in what is going on in Frisby v2.0. What would help: 
- add issues to the github what is being done, what needs to be done before considering this version usable etc.

**Comment by thim81** [comment]
@vlucas Could you perhaps tag the current v2.0 branch as an "alpha"? 
That way we can use it in a more controlled manner in packages.json and also allows to provide feedback per alpha tag release?

**Comment by markBarretto** [comment]
@vlucas, how would you add headers to a post request in v2.0?

**Comment by vasergen** [comment]
Hi, what the status of this version 2?

**Comment by mike123vn** [comment]
it died alreadly

**Comment by vlucas** [comment]
v2 is not dead... just stalled. I will look at tagging the branch as "alpha" so people can at least play around with it a bit more easily.

**Comment by vlucas** [comment]
I just tagged an alpha release:

`npm install frisby@2.0.0-alpha.1`

Please play with it and file an issue here if you find anything broken or unexpected.

**Comment by vlucas** [comment]
v2 is here, and has been moved to the `master` branch. Closing this ticket now.

**Comment by lenosky** [comment]
Hi!
I'm posting some data and need to get the response, @vlucas could you please point me at how can I do that with the v2?

**Comment by vlucas** [comment]
The docs are at: https://www.frisbyjs.com/http.html
